### PR TITLE
Support for D-Bus file descriptors

### LINKED ISSUE #4380: File descriptor support in dbus-json3
In order to call the full systemd DBus API we need file descriptor support. Here's how I imagined it working, although I'm sure there's other options.
1. When we receive a file descriptor in a DBus message, we keep that file descriptor open, until the dbus-json3 channel is closed.
2. In the marshalling of the message to send to the front end, substitute a unique string. Register that unique string as an "internal address" in cockpitchannel.c or similar.
3. Allow a new stream channel to be opened with { "internal": "uniquestring" }. The implementation here, dups the fd. It does not close the original.
4. When the DBus channel closes, the unique string is unregistered as an internal address.

Yes, in theory multiple channels can be opened with the same unique string. This has similar properties to file descriptors used in process.

But most importantly the above behavior, including dup'ing resolves races between frontend and backend knowing whether an FD is open, reused, and still represents what's expected.

### PR DESCRIPTION
**Pull request by larskarlitski**
This is based on @stefwalter's initial implementation at #4883. I've rebased it, fixed compile errors, and started to implement the missing bits. Still very much work in progress, but I'm opening this pull request to start the discussion again.

This is basic support for file descriptors passed over DBus. Several APIs like to include stuff like this in their APIs.
- [x] Initial implementation
- [x] Figure out possible fd exaustion until DBus channel is closed.
- [x] Figure out a test case.
- [ ] An actual use case.
- [x] Documentation
- [x] Unit tests, and fix bugs

Fixes #4380

### CODE REVIEW

**Comment by larskarlitski** [comment]
I think this is ready for a first review.

I've moved the internal channel logic into `CockpitPipeChannel`, because that's the one that deals with fds instead of addresses.

I've also added a small test case which opens a pipe and passes the reader to cockpit via a D-Bus message.

I've also documented that cockpit can now receive (but not sent) file descriptors over D-Bus. Do you think this warrants an extra section in the docs?

Does anyone have ideas for an actual use case that's small enough in scope to implement before 7.4?

File: doc/guide/cockpit-dbus.xml [hunk 1]
```diff
@@ -86,6 +86,13 @@
         <listitem><para>A javascript plain object with the <code>"t"</code> property set to a DBus type string,
             and the <code>"v"</code> property set to a value.</para></listitem>
       </varlistentry>
+      <varlistentry>
+        <term><code>HANDLE 'h'</code></term>
+        <listitem><para>A javascript object that describes a cockpit channel which represents the
+            passed file descriptor. Pass it to <link linkend="cockpit-channels-channel">
```
Comment (martinpitt) [inline doc/guide/cockpit-dbus.xml]: Could this be more specific about how its members look like?

File: src/bridge/cockpitpipechannel.c [hunk 2]
```diff
@@ -66,6 +66,21 @@ typedef struct {
 
 G_DEFINE_TYPE (CockpitPipeChannel, cockpit_pipe_channel, COCKPIT_TYPE_CHANNEL);
 
+GHashTable *internal_fds;
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: How does this simple hash table keep track of which fds are for reading and which for writing? Or do we only support receiving read fds for now?

File: src/base1/test-dbus-common.js [hunk 3]
```diff
@@ -1130,4 +1130,28 @@ function dbus_track_tests(channel_options, bus_name) {
                     });
             });
     });
+
+    QUnit.asyncTest("receive fd", function() {
```
Comment (martinpitt) [inline src/base1/test-dbus-common.js]: This tests receiving a read fd. It would be good to also have a test for getting a write fd, unless this shouldn't be supported (for now) -- but then that should be documented.

File: src/bridge/cockpitpipechannel.c [hunk 4]
```diff
@@ -384,6 +407,10 @@ cockpit_pipe_channel_prepare (CockpitChannel *channel)
       else
         self->pipe = cockpit_pipe_spawn ((const gchar **)argv, (const gchar **)env, dir, flags);
     }
+  else if (internal && lookup_internal_fd (internal, &fd))
+    {
+      self->pipe = cockpit_pipe_new (internal, fd, fd);
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: This specifies the same fd for both reading and writing, which is error prone and misleading. Unix pipes are unidirectional, and the sender and receiver need to agree which one gets passed and which one the sender keeps. This is related to my question above how we keep track of reading and writing fds.

File: src/bridge/cockpitpipechannel.c [hunk 5]
```diff
@@ -483,3 +510,46 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
...
+  gint fd = GPOINTER_TO_INT (data);
+
+  close (fd);
+}
+
+const gchar *
+cockpit_pipe_channel_add_internal_fd (gint fd)
+{
+  /* We are not multi-threaded. Also don't make this look like normal fd numbers */
+  static guint64 unique = 911111;
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: It seems to me that the keys of that GHashTable are mostly arbitrary and opaque, and we don't actually need them for anything? I must say I don't like these magic numbers. Would it be possible to use it as a set (`g_hash_table_add()` and `g_hash_table_contains()` and store the fds directly as keys, and use the fd as-is in the ID? Or can it happen that an fd gets closed and reused, but the corresponding cockpit channel doesn't get closed? (That would be an internal inconsistency and most likely lead to errors anyway?)

File: src/bridge/cockpitpipechannel.c [hunk 6]
```diff
@@ -483,3 +510,46 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
...
+
+  gchar *id;
+
+  if (!internal_fds)
+    internal_fds = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, internal_fd_free);
+
+  id = g_strdup_printf ("file-descriptor-%" G_GUINT64_FORMAT, unique++);
+  g_hash_table_replace (internal_fds, id, GINT_TO_POINTER (fd));
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: This comment would be moot when moving to a set,  but I think you should use `g_assert (g_hash_table_add ())` here -- like this we don't ever expect an ID to be already present.

File: src/bridge/cockpitpipechannel.c [hunk 7]
```diff
@@ -483,3 +510,46 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
...
+    return FALSE;
+
+  if (!g_hash_table_remove (internal_fds, id))
+    return FALSE;
+
+  if (g_hash_table_size (internal_fds) == 0)
+    {
+      g_hash_table_destroy (internal_fds);
+      internal_fds = NULL;
+    }
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: I'm not sure if that's worth the overhead -- if you keep adding and removing fds, you would always construct and delete the hash table. Seems easier and more efficient to just keep it around?

File: src/bridge/cockpitdbusjson.c [hunk 8]
```diff
@@ -691,7 +709,63 @@ build_json_dictionary (const GVariantType *entry_type,
...
+      node = json_node_new (JSON_NODE_OBJECT);
+
+      /* Add a new internal channel name for this file descriptor */
+      id = cockpit_pipe_channel_add_internal_fd (fd);
+      g_queue_push_tail (context->fdids, (gpointer)id);
+
+      /* And only keep the last N ready for opening channels */
+      while (g_queue_get_length (context->fdids) > MAX_RECEIVED_DBUS_FDS)
+        {
+          old = (const gchar *)g_queue_pop_head (context->fdids);
```
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: It seems a bit odd to silently kill the oldest fd instead of rejecting to receive a new one and returning an error instead? This feels much harder to handle from clients than checking for an error on sending. At least this dropping should cause a warning somewhere.

File: src/bridge/cockpitdbusjson.c [hunk 9]
```diff
@@ -2876,14 +2962,19 @@ cockpit_dbus_json_dispose (GObject *object)
...
-      g_object_run_dispose (G_OBJECT (peer->cache));
-      g_object_unref (peer->cache);
+
+      if (peer->cache)
+        {
+          g_signal_handler_disconnect (peer->cache, peer->meta_sig);
+          g_signal_handler_disconnect (peer->cache, peer->update_sig);
+          g_object_run_dispose (G_OBJECT (peer->cache));
+          g_object_unref (peer->cache);
+        }
```
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: This should be a separate commit. (Not important, though)

File: src/base1/test-dbus-common.js [hunk 10]
```diff
@@ -1130,4 +1130,28 @@ function dbus_track_tests(channel_options, bus_name) {
...
+                var fd = reply[0];
+                assert.equal(typeof(fd.internal), 'string');
+                assert.equal(fd.payload, 'stream');
+
+                var channel = cockpit.channel(fd);
+                QUnit.stop();
+                channel.onmessage = function (event, data) {
+                    assert.equal(data, 'Hello, fd');
+                    channel.close();
+                    QUnit.start();
```
Comment (larskarlitski) [inline src/base1/test-dbus-common.js]: Yes, start/stop are refcounted. We need this here because the `done()` path kicks off another async call to wait for the first message. We still need it in `always()` in case the dbus call fails. I'll add a comment.

File: doc/guide/cockpit-dbus.xml (hunk 1 above)
Comment (larskarlitski) [inline doc/guide/cockpit-dbus.xml]: I left it intentionally opaque, because the channel is openend via an internal id which is useless anyway. I'll add that this is always a stream channel.

File: src/bridge/cockpitpipechannel.c (hunk 4 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: Hm, good point. Is there any way to know if the fd we get from D-Bus is read- or write-only? The pipe is only for the test, but I guess this could happen in a practical use case as well.

File: src/bridge/cockpitpipechannel.c (hunk 5 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: We need a unique string to refer to the channel from cockpit. Maybe a random string that doesn't look like it *could* be a fd is even better?

File: src/bridge/cockpitpipechannel.c (hunk 7 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: Indeed :)

File: src/bridge/cockpitdbusjson.c (hunk 8 above)
Comment (larskarlitski) [inline src/bridge/cockpitdbusjson.c]: Client code can decide to ignore a passed fd. So we have to keep the fd open for  a while, so that the cockpit channel can be created if a client does not ignore it. The problem is when to close fds that were ignored. We decided to just keep the last N fds open.

This is why we can't warn about it, because many of the instances are just normal operation. If the limit turns out too low, the client code will get an error when it establishes the channel.

File: doc/guide/cockpit-dbus.xml (hunk 1 above)
Comment (martinpitt) [inline doc/guide/cockpit-dbus.xml]: Good point about the opacity, so fine as it is.

File: src/bridge/cockpitpipechannel.c (hunk 4 above)
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: The only way known to me is to check if `fdopen()` with a  particular mode returns `EINVAL` or succeeds. But I guess I may have misunderstood this -- so you are saying that cockpit itself should not touch/use the fd at all, just pass it between the service and the client? That would actually make a lot of sense, I just got misled when reviewing this.

Do I understand this alright -- with this, when the client opens a `cockpit.channel()` on this handle, it's still the client's responsibility to know whether it can read and/or write into this? If so, then we should really have a test case that ensures that cockpit doesn't touch the fds themselves -- i. e. trying to write to a read fd throws a sensible error, and we should have a test case for the write end of a pipe only to ensure that this direction works as well.

File: src/bridge/cockpitpipechannel.c (hunk 5 above)
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: So an fd isn't unique enough? Once a cockpit channel gets closed because the underlying fd closes, a new one must never use the same name again, even if the next passed fd has the same number than the previously closed one?

File: src/bridge/cockpitdbusjson.c (hunk 8 above)
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: Makes sense, thanks!

File: src/bridge/cockpitpipechannel.c (hunk 4 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: > Do I understand this alright

Yep, that's exactly it. I'll add the test cases you suggested.

File: src/bridge/cockpitpipechannel.c (hunk 5 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: No, the fd might be closed and reused before the client calls `cockpit.channel()`, which would give it the wrong fd.

Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: s/wrong fd/wrong id/ I suppose. Understood, so let's not dwell on this further -- it's just an opaque ID, so we can change it later on if we want.

File: src/bridge/cockpitpipechannel.c (hunk 2 above)
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: Based on the discussion below, this comment is moot.

File: src/bridge/cockpitpipechannel.c (hunk 6 above)
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: I'd actually do the assertion here -- as we've discussed above, we do want the IDs  to be unique even if the fds get reused, so this prevents falling into a trap in the future when someone "simplifies" the ID housekeeping.

**Comment by larskarlitski** [comment]
Update: implemented @martinpitt's suggestions. Thanks a lot!

File: src/common/mock-service.c [hunk 11]
```diff
@@ -454,6 +456,68 @@ on_tell_me_your_name (TestFrobber *frobber,
...
+test_fd_pipe_readable (gint fd,
+                       GIOCondition condition,
+                       gpointer user_data)
+{
+  const char *expected = "Hello, fd";
+  char buffer[100];
+
+  g_assert (condition == G_IO_IN);
+
+  read (fd, buffer, 100);
```
Comment (dperpeet) [inline src/common/mock-service.c]: semaphore complains here while checking:
```
src/common/mock-service.c: In function ‘test_fd_pipe_readable’:
src/common/mock-service.c:469:8: error: ignoring return value of ‘read’, declared with attribute warn_unused_result [-Werror=unused-result]
   read (fd, buffer, 100);
        ^
```
Even in a mock function it's probably good practice to check this.

**Comment by dperpeet** [comment]
I didn't review anything besides the Semaphore error.

File: src/common/mock-service.c (hunk 11 above)
Comment (larskarlitski) [inline src/common/mock-service.c]: Thanks for the catch. I've added a similar check like I have for `write`. I think we can safely assume that we always get all data at once since this is a pipe and very little data.

File: src/bridge/cockpitpipechannel.c [hunk 12]
```diff
@@ -384,6 +407,10 @@ cockpit_pipe_channel_prepare (CockpitChannel *channel)
       else
         self->pipe = cockpit_pipe_spawn ((const gchar **)argv, (const gchar **)env, dir, flags);
     }
+  else if (internal && lookup_internal_fd (internal, &fd))
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: So the following line claims the fd, however lookup_internal_fd doesn't steal it from the hashtable or otherwise transfer ownership.

I imagine that such fd channels are open-once channels, after which the identifier becomes invalid? Or did you imagine them to be able to open many channels to the same fd? This is obviously a corner case, but either way needs to be made consistent.

File: src/bridge/cockpitpipechannel.c [hunk 13]
```diff
@@ -66,6 +66,21 @@ typedef struct {
 
 G_DEFINE_TYPE (CockpitPipeChannel, cockpit_pipe_channel, COCKPIT_TYPE_CHANNEL);
 
+GHashTable *internal_fds;
+
+static gboolean
+lookup_internal_fd (const gchar *name, gint *fdp)
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: Arguments should be on different lines.

File: src/bridge/cockpitpipechannel.c [hunk 14]
```diff
@@ -483,3 +510,43 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
...
+  /* We are not multi-threaded. Also don't make this look like normal fd numbers */
+  static guint64 unique = 911111;
+  gboolean inserted;
+
+  gchar *id;
+
+  if (!internal_fds)
+    internal_fds = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, internal_fd_free);
+
+  id = g_strdup_printf ("file-descriptor-%" G_GUINT64_FORMAT, unique++);
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: This name is a bit misleading. For folks debugging this, they'll think that the file descriptor N is being opened. Worth choosing another name like 'internal-stream-XXX' or something like that? If you don't think so, I'm fine with keeping it as it is. Just consider it.

File: src/bridge/cockpitpipechannel.c (hunk 12 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: I didn't bother because you wrote this

> Yes, in theory multiple channels can be opened with the same unique string. This has similar properties to file descriptors used in process.

in bug #4380. Changed it to remove the channel and added a test.

File: src/bridge/cockpitpipechannel.c (hunk 14 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: Changed it.

**Comment by larskarlitski** [comment]
Update: address issues raised by @stefwalter. Thanks!

File: src/bridge/cockpitpipechannel.c (hunk 12 above)
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: Well it would have been fine if you had dup()'d it ... then the channel could have been opened more than once. But I think this makes sense, just have it be opened once.
